src/database/database-postgresql.cpp

```diff
@@ -812,5 +813,210 @@ void AuthDatabasePostgreSQL::writePrivileges(const AuthEntry &authEntry)
+					"UPDATE SET value = $3::bytea");
+	}
+	prepareStatement("remove",
+		"DELETE FROM mod_storage WHERE modname = $1 AND key = $2::bytea");
+	prepareStatement("remove_all",
+		"DELETE FROM mod_storage WHERE modname = $1");
+	prepareStatement("list",
+		"SELECT modname FROM mod_storage GROUP BY modname");
+}
+
+bool ModMetadataDatabasePostgreSQL::getModEntries(const std::string &modname, StringMap *storage)
+{
+	verifyDatabase();
+
+	const void *args[] = { modname.c_str() };
+	const int argLen[] = { -1 };
+	const int argFmt[] = { 0 };
+	PGresult *results = execPrepared("get_all", ARRLEN(args),
+			args, argLen, argFmt, false);
+
+	int numrows = PQntuples(results);
+
+	for (int row = 0; row < numrows; ++row) {
+		std::string key(PQgetvalue(results, row, 0), PQgetlength(results, row, 0));
+		std::string value(PQgetvalue(results, row, 1), PQgetlength(results, row, 1));
+		storage->emplace(std::move(key), std::move(value));
+	}
+
+	PQclear(results);
+
+	return true;
+}
+
+bool ModMetadataDatabasePostgreSQL::getModKeys(const std::string &modname,
+		std::vector<std::string> *storage)
+{
+	verifyDatabase();
+
+	const void *args[] = { modname.c_str() };
+	const int argLen[] = { -1 };
+	const int argFmt[] = { 0 };
+	PGresult *results = execPrepared("get_all_keys", ARRLEN(args),
+			args, argLen, argFmt, false);
+
+	int numrows = PQntuples(results);
+
+	for (int row = 0; row < numrows; ++row)
+		storage->emplace_back(PQgetvalue(results, row, 0), PQgetlength(results, row, 0));
+
+	PQclear(results);
+
+	return true;
+}
+
+bool ModMetadataDatabasePostgreSQL::getModEntry(const std::string &modname,
+	const std::string &key, std::string *value)
+{
+	verifyDatabase();
+
+	const void *args[] = { modname.c_str(), key.c_str() };
+	const int argLen[] = { -1, (int)MYMIN(key.size(), (size_t)INT_MAX) };
+	const int argFmt[] = { 0, 1 };
+	PGresult *results = execPrepared("get", ARRLEN(args), args, argLen, argFmt, false);
+
+	int numrows = PQntuples(results);
+	bool found = numrows > 0;
+
+	if (found)
+		value->assign(PQgetvalue(results, 0, 0), PQgetlength(results, 0, 0));
+
+	PQclear(results);
+
+	return found;
+}
+
+bool ModMetadataDatabasePostgreSQL::hasModEntry(const std::string &modname,
+		const std::string &key)
+{
+	verifyDatabase();
+
+	const void *args[] = { modname.c_str(), key.c_str() };
+	const int argLen[] = { -1, (int)MYMIN(key.size(), (size_t)INT_MAX) };
+	const int argFmt[] = { 0, 1 };
+	PGresult *results = execPrepared("has", ARRLEN(args), args, argLen, argFmt, false);
+
+	int numrows = PQntuples(results);
+	bool found = numrows > 0;
+
+	PQclear(results);
+
+	return found;
+}
+
+bool ModMetadataDatabasePostgreSQL::setModEntry(const std::string &modname,
+	const std::string &key, const std::string &value)
+{
+	verifyDatabase();
+
+	const void *args[] = { modname.c_str(), key.c_str(), value.c_str() };
+	const int argLen[] = {
+		-1,
+		(int)MYMIN(key.size(), (size_t)INT_MAX),
+		(int)MYMIN(value.size(), (size_t)INT_MAX),
```

Comment:
suggestion:
```cpp
template<typename T, typename = std::enable_if_t<std::is_unsigned<T>::value>>
int clamp_int(T n) {
    return static_cast<int>(MYMIN(n, INT_MAX));
}
```

Reply:
I have simplified the code a bit so that I don't think a helper function is necessary.